source/sigv4.c

```diff
@@ -947,17 +1236,515 @@ static SigV4Status_t verifySigV4Parameters( const SigV4Parameters_t * pParams )
+                                                   pInput, inputLen );
+    }
+
+    if( hashStatus == 0 )
+    {
+        hashStatus = pCryptoInterface->hashFinal( pCryptoInterface->pHashContext,
+                                                  pOutput, outputLen );
+    }
+
+    return hashStatus;
+}
+
+/*-----------------------------------------------------------*/
+
+static SigV4Status_t completeHashAndHexEncode( const char * pInput,
+                                               size_t inputLen,
+                                               char * pOutput,
+                                               size_t * pOutputLen,
+                                               const SigV4CryptoInterface_t * pCryptoInterface )
+{
+    SigV4Status_t returnStatus = SigV4Success;
+    /* Used to store the hash of the request payload. */
+    char hashedPayload[ SIGV4_HASH_MAX_DIGEST_LENGTH ];
+    SigV4String_t originalHash;
+    SigV4String_t hexEncodedHash;
+
+    assert( pOutput != NULL );
+    assert( pOutputLen != NULL );
+    assert( pCryptoInterface != NULL );
+
+    originalHash.pData = hashedPayload;
+    originalHash.dataLen = pCryptoInterface->hashDigestLen;
+    hexEncodedHash.pData = pOutput;
+    hexEncodedHash.dataLen = *pOutputLen;
+
+    if( completeHash( pInput,
+                      inputLen,
+                      hashedPayload,
+                      pCryptoInterface->hashDigestLen,
+                      pCryptoInterface ) != 0 )
+    {
+        returnStatus = SigV4HashError;
+    }
+
+    if( returnStatus == SigV4Success )
+    {
+        /* Hex-encode the request payload. */
+        returnStatus = lowercaseHexEncode( &originalHash,
+                                           &hexEncodedHash );
+    }
+
+    if( returnStatus == SigV4Success )
+    {
+        *pOutputLen = hexEncodedHash.dataLen;
     }
 
     return returnStatus;
 }
 
+int32_t hmacKey( HmacContext_t * pHmacContext,
+                 const char * pKey,
+                 size_t keyLen )
+{
+    int32_t returnStatus = 0;
+    const SigV4CryptoInterface_t * pCryptoInterface = NULL;
+
+    assert( pHmacContext != NULL );
+    assert( pHmacContext->key != NULL );
+    assert( pHmacContext->pCryptoInterface != NULL );
+    assert( pHmacContext->pCryptoInterface->hashInit != NULL );
+    assert( pHmacContext->pCryptoInterface->hashUpdate != NULL );
+    assert( pHmacContext->pCryptoInterface->hashFinal != NULL );
+
+    pCryptoInterface = pHmacContext->pCryptoInterface;
+
+    if( pHmacContext->keyLen + keyLen <= pCryptoInterface->hashBlockLen )
+    {
+        /* The key fits into the block so just append it. */
+        ( void ) memcpy( pHmacContext->key + pHmacContext->keyLen, pKey, keyLen );
+    }
+    else
+    {
+        /* Initialize the hash context and hash existing key data. */
+        if( pHmacContext->keyLen <= pCryptoInterface->hashBlockLen )
+        {
+            returnStatus = pCryptoInterface->hashInit( pCryptoInterface->pHashContext );
```

Comment:
Some comments would go a long way for explaining why this can only be called once per hmac calculation